In [1]:
import numpy as np
import pandas as pd
from keras.utils import to_categorical

In [4]:
dataset = pd.read_csv('/content/Iris.csv')
dataset

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

In [5]:
X = dataset.iloc[:,1:5].values
y = dataset.iloc[:,5].values

from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
y1 = encoder.fit_transform(y)

Y = pd.get_dummies(y1).values
print(Y)

[[1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 

In [9]:
from sklearn.model_selection import train_test_split

X_train,X_test, y1_train, y1_test = train_test_split(X,Y,test_size=0.2)
#print(X_test)

In [37]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD, Adam

model = Sequential()

model.add(Dense(10,input_shape=(4,),activation='tanh'))
model.add(Dense(8, activation='tanh'))
model.add(Dense(6, activation='tanh'))
model.add(Dense(3, activation='softmax'))

model.compile(SGD(lr=0.01), 'mean_squared_error', metrics=['accuracy'])

model.summary()

Model: "sequential_15"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_60 (Dense)             (None, 10)                50        
_________________________________________________________________
dense_61 (Dense)             (None, 8)                 88        
_________________________________________________________________
dense_62 (Dense)             (None, 6)                 54        
_________________________________________________________________
dense_63 (Dense)             (None, 3)                 21        
Total params: 213
Trainable params: 213
Non-trainable params: 0
_________________________________________________________________


In [39]:
model.fit(X_train,y1_train,epochs =  100)
y_pred = model.predict(X_test)

Epoch 1/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1654 - accuracy: 0.6917
Epoch 2/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1645 - accuracy: 0.6917
Epoch 3/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1637 - accuracy: 0.6917
Epoch 4/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1629 - accuracy: 0.6917
Epoch 5/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1622 - accuracy: 0.6917
Epoch 6/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1614 - accuracy: 0.6917
Epoch 7/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1606 - accuracy: 0.6917
Epoch 8/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1598 - accuracy: 0.6917
Epoch 9/100
4/4 [==============================] - 0s 4ms/step - loss: 0.1590 - accuracy: 0.6917
Epoch 10/100
4/4 [==============================] - 0s 2ms/step - loss: 0.1582 - accuracy: 0.6917
Epoch 11/100
4/4 [===========

**SGD with Mean Squared Error**

|LR|Loss|Accuracy|
|---|---| ---|
|0.01|0.2246|0.35|
|0.1|0.0446|0.92|
|1|0.0456|0.92|

**SGD with Categorical Cross Entropy**

|LR|Loss|Accuracy|
|---|---|---|
|0.01|0.3480|0.95|
|0.1|0.2123|0.91|
|1|1.1637|0.29|

**Adam with Mean Squared Error**

|LR|Loss|Accuracy|
|---|---|---|
|0.01|0.0151|0.96|
|0.1|0.1104|0.69|
|1|0.4333|0.35|

**Adam with Categorical Cross Entropy**

|LR|Loss|Accuracy|
|---|---|---|
|0.01|0.0877|0.96|
|0.1|0.4640|0.65|
|1|1.357|0.30|

From overall observation, we can say that LR=0.01 get highest accuracy generally but then it depends on how the data is split. As when we run this model for same set of parameters, it gets different accuracy. Also, Adam performs better than SGD. 